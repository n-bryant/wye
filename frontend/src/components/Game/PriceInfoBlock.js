import React from "react";
import PropTypes from "prop-types";

import { withStyles } from "@material-ui/core/styles";
import { withWidth } from "@material-ui/core";
import createClassNameHelper from "@n_bryant/classnames-helper";
import JSS_CLASS_NAME_PREFIX from "../../../lib/classNamePrefix";

import { useRouter } from "next/router";

import Icon from "@mdi/react";
import { mdiPlusMinus } from "@mdi/js";

import Button from "@material-ui/core/Button";
import Typography from "@material-ui/core/Typography";

import PriceWidget from "./PriceWidget";
import styles from "./PriceInfoBlock.styles";

/**
 * renders an info block for a game's price and review details
 */
export const PriceInfoBlock = props => {
  const classnames = PriceInfoBlock.classnames(props);
  const {
    data: { price, metacritic },
    width
  } = props;
  const router = useRouter();

  const hasSmScreen = ["xs", "sm"].some(val => val === width);

  return (
    <div
      className={classnames.root({
        withSmScreen: hasSmScreen
      })}
    >
      <div
        className={classnames.element("priceContainer", {
          withSmScreen: hasSmScreen
        })}
      >
        <PriceWidget
          data={price}
          className={classnames.element("priceWidget")}
        />
        <a
          className={classnames.element("link")}
          href={`https://store.steampowered.com/app/${router.query.id}`}
          target="_blank"
        >
          <Button className={classnames.element("storeButton")}>
            <Typography variant="body1">Visit Store Page</Typography>
          </Button>
        </a>
      </div>
      <div
        className={classnames.element("metacriticContainer", {
          withSmScreen: hasSmScreen
        })}
      >
        <div className={classnames.element("metacriticTitle")}>
          <Icon className={classnames.element("icon")} path={mdiPlusMinus} />
          <Typography variant="h1">Metacritic Score: </Typography>
        </div>
        <div>
          <Typography
            className={classnames.element("metacriticScore", {
              negative: metacritic.score < 50
            })}
            variant="body1"
          >
            {metacritic.score > 0 ? (
              <a
                className={classnames.element("link")}
                href={metacritic.reviewsPageUrl}
                target="_blank"
              >
                {`${metacritic.score}%`}
              </a>
            ) : (
              "n/a"
            )}
          </Typography>
        </div>
      </div>
    </div>
  );
};
PriceInfoBlock.classnames = createClassNameHelper(
  `${JSS_CLASS_NAME_PREFIX}PriceInfoBlock`
);
PriceInfoBlock.propTypes = {
  // styles to apply
  classes: PropTypes.shape({
    root: PropTypes.string,
    withSmScreen: PropTypes.string,
    priceContainer: PropTypes.string,
    priceContainerWithSmScreen: PropTypes.string,
    priceWidget: PropTypes.string,
    storeButton: PropTypes.string,
    link: PropTypes.string,
    metacriticContainer: PropTypes.string,
    metacriticContainerWithSmScreen: PropTypes.string,
    metacriticTitle: PropTypes.string,
    metacriticScore: PropTypes.string,
    metacriticScoreNegative: PropTypes.string,
    icon: PropTypes.string
  }),
  // data for the game relating to price and reviews
  data: PropTypes.shape({
    price: PropTypes.shape({
      freeToPlay: PropTypes.bool,
      onSale: PropTypes.bool,
      discountPercentage: PropTypes.number,
      initialFormatted: PropTypes.string,
      finalFormatted: PropTypes.string
    }),
    metacritic: PropTypes.shape({
      score: PropTypes.number,
      reviewsPageUrl: PropTypes.string
    })
  }),
  // material-ui breakpoint width
  width: PropTypes.string
};
PriceInfoBlock.defaultProps = {
  classes: {}
};

export default withWidth()(withStyles(styles)(PriceInfoBlock));
