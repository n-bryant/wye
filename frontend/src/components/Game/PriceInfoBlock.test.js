import { PriceInfoBlock } from "./PriceInfoBlock";

jest.mock("next/router", () => ({
  useRouter: () => ({
    query: {
      id: "1"
    }
  })
}));

describe("PriceInfoBlock", () => {
  const props = {
    classes: {
      root: "root",
      withSmScreen: "withSmScreen",
      priceContainer: "priceContainer",
      priceContainerWithSmScreen: "priceContainerWithSmScreen",
      priceWidget: "priceWidget",
      storeButton: "storeButton",
      link: "link",
      metacriticContainer: "metacriticContainer",
      metacriticContainerWithSmScreen: "metacriticContainerWithSmScreen",
      metacriticTitle: "metacriticTitle",
      metacriticScore: "metacriticScore",
      metacriticScoreNegative: "metacriticScoreNegative",
      icon: "icon"
    },
    data: {
      price: {
        freeToPlay: true,
        onSale: true,
        discountPercentage: 50,
        initialFormatted: "$10.00",
        finalFormatted: "$5.00"
      },
      metacritic: {
        score: 0,
        reviewsPageUrl: "url"
      }
    }
  };

  it("should render successfully", () => {
    const wrapper = shallow(<PriceInfoBlock {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should render a link with the metacritic score and way to navigate to the reviews page if the metacritic score is above 0", () => {
    const propsWithNonZeroScore = {
      ...props,
      data: {
        ...props.data,
        metacritic: {
          score: 99,
          url: "url"
        }
      }
    };
    const wrapperWithNonZeroScore = shallow(
      <PriceInfoBlock {...propsWithNonZeroScore} />
    );
    const metacriticScore = wrapperWithNonZeroScore.findWhere(n =>
      n.hasClass(props.classes.metacriticScore)
    );
    expect(metacriticScore.length).toBe(1);
    expect(toJson(metacriticScore)).toMatchSnapshot();
  });

  it("should set negative styling to the metacritic score if the score is less than 50", () => {
    const wrapper = shallow(<PriceInfoBlock {...props} />);
    expect(
      wrapper
        .findWhere(n => n.hasClass(props.classes.metacriticScore))
        .hasClass(props.classes.metacriticScoreNegative)
    ).toBeTruthy();
  });
});
