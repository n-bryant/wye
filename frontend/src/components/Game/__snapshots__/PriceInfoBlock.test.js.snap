// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PriceInfoBlock should render a link with the metacritic score and way to navigate to the reviews page if the metacritic score is above 0 1`] = `
<WithStyles(ForwardRef(Typography))
  className="wye-PriceInfoBlock__metacriticScore metacriticScore"
  variant="body1"
>
  <a
    className="wye-PriceInfoBlock__link link"
    target="_blank"
  >
    99%
  </a>
</WithStyles(ForwardRef(Typography))>
`;

exports[`PriceInfoBlock should render successfully 1`] = `
<div
  className="wye-PriceInfoBlock root"
>
  <div
    className="wye-PriceInfoBlock__priceContainer priceContainer"
  >
    <WithStyles(PriceWidget)
      className="wye-PriceInfoBlock__priceWidget priceWidget"
      classes={Object {}}
      data={
        Object {
          "discountPercentage": 50,
          "finalFormatted": "$5.00",
          "freeToPlay": true,
          "initialFormatted": "$10.00",
          "onSale": true,
        }
      }
      skinny={false}
    />
    <a
      className="wye-PriceInfoBlock__link link"
      href="https://store.steampowered.com/app/1"
      target="_blank"
    >
      <WithStyles(ForwardRef(Button))
        className="wye-PriceInfoBlock__storeButton storeButton"
      >
        <WithStyles(ForwardRef(Typography))
          variant="body1"
        >
          Visit Store Page
        </WithStyles(ForwardRef(Typography))>
      </WithStyles(ForwardRef(Button))>
    </a>
  </div>
  <div
    className="wye-PriceInfoBlock__metacriticContainer metacriticContainer"
  >
    <div
      className="wye-PriceInfoBlock__metacriticTitle metacriticTitle"
    >
      <Icon
        className="wye-PriceInfoBlock__icon icon"
        color={null}
        horizontal={false}
        path="M11,4V9H6V11H11V16H13V11H18V9H13V4H11M6,18V20H18V18H6Z"
        rotate={0}
        size={null}
        spin={false}
        vertical={false}
      />
      <WithStyles(ForwardRef(Typography))
        variant="h1"
      >
        Metacritic Score: 
      </WithStyles(ForwardRef(Typography))>
    </div>
    <div>
      <WithStyles(ForwardRef(Typography))
        className="wye-PriceInfoBlock__metacriticScore wye-PriceInfoBlock__metacriticScore--negative metacriticScore metacriticScoreNegative"
        variant="body1"
      >
        n/a
      </WithStyles(ForwardRef(Typography))>
    </div>
  </div>
</div>
`;
